yt_dlp/extractor/ximalaya.py

```diff
@@ -71,8 +72,157 @@ class XimalayaIE(XimalayaBaseIE):
                 'like_count': int,
             },
         },
+        {
+            # VIP-restricted audio
+            'url': 'https://www.ximalaya.com/sound/562111701',
+            'only_matching': True,
+        },
     ]
 
+    @staticmethod
+    def _decrypt_filename(audio_info):
+        seed = float_or_none(audio_info['seed'])
+        file_id = audio_info['fileId']
+        cgstr = ''
+        key = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ/\\:._-1234567890'
+        for _ in range(len(key)):
+            j = 211 * seed + 30031
+            seed = float_or_none(int_or_none(j) % 65536)
+            ran = seed / float_or_none(65536)
+            r = int_or_none(ran * float_or_none(len(key)))
+            cgstr += key[r]
+            key = key.replace(key[r], '')
+        strs = file_id.split('*')
+        filename = ''
+        for n in range(len(strs) - 1):
+            if strs[n] != '':
+                index = int_or_none(strs[n])
+                filename += cgstr[index]
+        if filename[0] != '/':
+            filename = '/' + filename
+        return filename
+
+    @staticmethod
+    def _decrypt_url_params(audio_info):
+        def char_code_at(s: str, n: int) -> int:
+            return ord(s[n]) if n < len(s) else 0
+
+        def decrypt(e: str, t: list) -> str:
```

Comment:
This is just rc4, right?
Maybe we can use some more commonly used names here.

Reply:
also: https://github.com/yt-dlp/yt-dlp/pull/10888/files#diff-f6f324eb09736a79244ee07009f4a3b05a67d7130c4a17749448c293f8210fe0R103

this should probably be a utils function, or better yet part of a proposed crypto module

Reply:
Yes, this is basically rc4. Regarding the proposal to separate the rc4 function into a common function, I suggest that we implement this in a new PR after merging #10832 and #10888.